cl/replication/tests: deflake test_with_restart

bharathv opened the pull request:
See individual commits.

Fixes: https://redpandadata.atlassian.net/browse/CORE-14672

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
cl/group_mirroring: fix missing broker_id in log line

bharathv commented:
/ci-repeat 1
dt-repeat=50
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_with_restart

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e74c8b28-5b23-4129-806e-22b53c586aa5-->
<!--TEST RESULTS-->
<!--070e655e-c3e9-470b-9652-323f718b653c-->
<!--TEST RESULTS-->
<!--88fe19f7-43d8-4a52-838d-e76de2b02831-->
<!--TEST RESULTS-->
<!--19b64f92-2d11-4142-8ab1-f2ac6a83f415-->
<!--TEST RESULTS-->
<!--95646467-fb74-4df9-8d8a-5277ec6fe8c6-->
<!--TEST RESULTS-->
<!--bc2668b1-8e6d-44ec-96b0-5062031a5241-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76687">build#76687</a></summary>


| test_class                    | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                  |
|:------------------------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_with_restart | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76687#019a9e5b-88d8-46ac-8426-03f412a79e4c | FLAKY         | 69/70    | upstream reliability is '99.71910112359551'. current run reliability is '95.23809523809523'. drift is 4.48101 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_with_restart |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76770">build#76770</a></summary>


| test_class                    | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                  |
|:------------------------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_with_restart | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76770#019aa4ce-a677-49dd-b9b5-d50a81c84696 | FLAKY         | 118/120  | upstream reliability is '99.75'. current run reliability is '90.9090909090909'. drift is 8.84091 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_with_restart |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77384">build#77384</a></summary>


| test_class                           | test_method                                  | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                               |
|:-------------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring               | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24c9-4d80-b6cc-1aa9770be6f8 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring          |
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring               | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb10-405a-b47e-c36f79b30dc9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring          |
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring               | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24ca-4a8f-b7c6-7505f47ec84f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring          |
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring               | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb12-4e6b-a1f4-a140a951d038 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring          |
| ShadowLinkingReplicationTests        | test_topic_delete                            | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb12-4e6b-a1f4-a140a951d038 | FLAKY         | 16/21    | upstream reliability is '93.75'. current run reliability is '76.19047619047619'. drift is 17.55952 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete                              |
| ConsumerOffsetsRecoveryTest          | test_consumer_offsets_partition_recovery     | {"force_offset_upload_failures": false}                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24c7-4fba-bd9e-a073508755ab | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryTest&test_method=test_consumer_offsets_partition_recovery         |
| ConsumerOffsetsRecoveryTest          | test_consumer_offsets_partition_recovery     | {"force_offset_upload_failures": false}                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb0e-4e6c-aae5-162306efff37 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryTest&test_method=test_consumer_offsets_partition_recovery         |
| ConsumerOffsetsRecoveryTest          | test_consumer_offsets_partition_recovery     | {"force_offset_upload_failures": true}                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24c9-4d80-b6cc-1aa9770be6f8 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryTest&test_method=test_consumer_offsets_partition_recovery         |
| ConsumerOffsetsRecoveryTest          | test_consumer_offsets_partition_recovery     | {"force_offset_upload_failures": true}                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb10-405a-b47e-c36f79b30dc9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryTest&test_method=test_consumer_offsets_partition_recovery         |
| ConsumerOffsetsRecoveryToolTest      | test_consumer_offsets_partition_count_change | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24ca-4a8f-b7c6-7505f47ec84f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryToolTest&test_method=test_consumer_offsets_partition_count_change |
| ConsumerOffsetsRecoveryToolTest      | test_consumer_offsets_partition_count_change | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb12-4e6b-a1f4-a140a951d038 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsRecoveryToolTest&test_method=test_consumer_offsets_partition_count_change |
| NodesDecommissioningTest             | test_decommissioning_rebalancing_node        | {"shutdown_decommissioned": true}                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24c7-4fba-bd9e-a073508755ab | FLAKY         | 16/21    | upstream reliability is '92.3507462686567'. current run reliability is '76.19047619047619'. drift is 16.16027 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node               |
| RedpandaNodeOperationsSmokeTest      | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": true}                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed19-fb0a-41d0-adb2-9cbcc3d8a764 | FLAKY         | 9/21     | upstream reliability is '86.11111111111111'. current run reliability is '42.857142857142854'. drift is 43.25397 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                     |
| WriteCachingFailureInjectionE2ETest  | test_crash_all                               | {"use_transactions": false}                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/77384#019aed14-24c6-4907-80ce-4da059ade838 | FLAKY         | 20/21    | upstream reliability is '88.67521367521367'. current run reliability is '95.23809523809523'. drift is -6.56288 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77449">build#77449</a></summary>


| test_class    | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                        |
|:--------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| ScalingUpTest | test_fast_node_addition | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77449#019af002-5f1e-4eba-887e-ace2b85628c4 | FLAKY         | 29/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0262, p0=0.5495, reject_threshold=0.0100. adj_baseline=0.0766, p1=0.3191, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77471">build#77471</a></summary>


| test_class                       | test_method                                  | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                         |
|:---------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests    | test_topic_delete                            | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af11d-4e4b-4fee-97cd-2961bd4772ad | FLAKY         | 8/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete                        |
| ShadowLinkingReplicationTests    | test_topic_delete                            | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-483b-4788-a5bd-41cac02ed7ac | FLAKY         | 14/21    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0728, p0=0.0024, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete                        |
| ConsumerOffsetsConsistencyTest   | test_flipping_leadership                     | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af11d-4e4b-4fee-97cd-2961bd4772ad | FLAKY         | 4/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsConsistencyTest&test_method=test_flipping_leadership                |
| ConsumerOffsetsConsistencyTest   | test_flipping_leadership                     | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-483a-48af-bb77-78d7f8310698 | FLAKY         | 3/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsConsistencyTest&test_method=test_flipping_leadership                |
| DatalakeDLQTest                  | test_dlq_table_for_mixed_records             | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af11d-4e47-44ed-996e-bee3197e8e20 | FLAKY         | 49/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0005, p0=0.0251, reject_threshold=0.0100. adj_baseline=0.0015, p1=0.9973, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_mixed_records                       |
| NodesDecommissioningTest         | test_decommissioning_rebalancing_node        | {"shutdown_decommissioned": true}                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-483a-48af-bb77-78d7f8310698 | FLAKY         | 16/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0896, p0=0.0982, reject_threshold=0.0100. adj_baseline=0.2455, p1=0.4330, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node         |
| NodesDecommissioningTest         | test_flipping_decommission_recommission      | {"cloud_topic": true, "node_is_alive": false}                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af11d-4e4b-4fee-97cd-2961bd4772ad | FLAKY         | 50/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0000, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_flipping_decommission_recommission       |
| PartitionMoveInterruption        | test_cancellations_interrupted_with_restarts | {"replication_factor": 3}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-4839-4ea2-9cfc-ce0cb191a923 | FLAKY         | 7/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMoveInterruption&test_method=test_cancellations_interrupted_with_restarts |
| RedpandaNodeOperationsSmokeTest  | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": false}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af11d-4e46-4544-9ed1-36f5e23530a1 | FLAKY         | 38/41    | Test is INCONCLUSIVE after retries.[Inconclusive result before max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-D%3A-inconclusive-result-after-max-retries-(run-fails)%5BinlineExtension%5D)(baseline=0.0115, p0=0.0773, reject_threshold=0.0100. adj_baseline=0.0341, p1=0.8449, trust_threshold=0.5000)                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test               |
| ScalingUpTest                    | test_fast_node_addition                      | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-4839-4ea2-9cfc-ce0cb191a923 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0261, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0763, p1=0.4520, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                                  |
| ScalingUpTest                    | test_moves_with_local_retention              | {"use_topic_property": true}                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-483b-4788-a5bd-41cac02ed7ac | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0253, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0740, p1=0.4637, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_moves_with_local_retention                          |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss                   | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/77471#019af120-4835-4b83-97dc-d4670e4190fb | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0515, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1466, p1=0.2050, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77540">build#77540</a></summary>


| test_class                    | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                  |
|:------------------------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_with_restart | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77540#019affa1-1829-42a0-9adf-6d6add4d58be | FLAKY         | 59/70    | upstream reliability is '100.0'. current run reliability is '62.06896551724138'. drift is 37.93103 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_with_restart |
| ShadowLinkingReplicationTests | test_with_restart | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77540#019affa1-182c-4c7a-aad5-67fb7f68ba40 | FLAKY         | 62/70    | upstream reliability is '100.0'. current run reliability is '68.0'. drift is 32.0 and the allowed drift is set to 50. The test should PASS                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_with_restart |
</details>
<!--/body-->

bharathv committed:
sl/tests: add ability to pass consumer properties

bharathv committed:
kgo: ability to set rebalance/session timeouts

bharathv committed:
cl/tests: run with aggressive rebalance/session timeouts

bharathv commented:
/ci-repeat 1
dt-repeat=50
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_with_restart

bharathv commented:
/ci-repeat 1
skip-redpanda-build
skip-units
dt-repeat=100
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_with_restart

bharathv committed:
kgo: make exactly once consumption configurable

No logical changes, the flag is true by default.

bharathv committed:
cl/tests: disable exactly_once in test_with_restart

The test does a rolling restart of all brokers. It runs with with
background auto commit (which runs every 5s), this is prone to failing
offset commits every once in a while and records may be consumed again.

bharathv commented:
/ci-repeat 1
dt-repeat=75
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_with_restart

dotnwat commented:
added @nvartolomei whose been poking around kgo recently and might be interested.

mmaslankaprv commented on tests/go/kgo-verifier/pkg/worker/verifier/validator_status.go:
```diff
@@ -91,7 +95,7 @@ func (cs *ValidatorStatus) ValidateRecord(r *kgo.Record, validRanges *TopicOffse
 				log.Warnf("Gap detected in consumed offsets. Expected %d, but got %d", expected, r.Offset)
 			}
-		} else {
+		} else if cs.exactlyOnceConsumption {
```
I am wondering if there is a way to validate that the message that is consumed more than once is the same as seen before. I do not remember if we store enough state in kgo-verifier to check that

bharathv replied:
I think we already have that out of the box because we persist the offset in the record payload and validate that on consumption https://github.com/redpanda-data/redpanda/blob/0127bc2826a9a91a5741817482c017e6ef305b57/tests/go/kgo-verifier/pkg/worker/verifier/validator_status.go#L116

nvartolomei commented:
@bharathv I already account for "at least once" consume mode when consumer group restarts at https://github.com/redpanda-data/redpanda/blob/b576a0f7633f864d668eb55557459de4e0fd8761/tests/go/kgo-verifier/pkg/worker/verifier/group_read_worker.go#L197-L204

First, why is this not enough?

bharathv commented:
> @bharathv I already account for "at least once" consume mode when consumer group restarts at
> 
> https://github.com/redpanda-data/redpanda/blob/b576a0f7633f864d668eb55557459de4e0fd8761/tests/go/kgo-verifier/pkg/worker/verifier/group_read_worker.go#L197-L204
> 
> First, why is this not enough?

oh, I didn't know about this check.  Now that I read it I don't understand how this is supposed to work.  What err is `consumerGroupReadInner` expected to return if there was a duplicate read? Group rebalance is an async operation, so an example where this doesn't work is

t0 - fetched until offset=50 for p0
t1 - async offset committer - committed until offset 50 for p0 
t2 - pollRecords() fetched until hwm (100) - waiting to fetch more (nothing to fetch)
t3 - Rebalance ran async and all partition assigments are revoked (will block offset commits)
t4 - Assigned again for p=0 at offset=50 (last committed) after syncGroup()
t5 - pollFetches() returns again from offset 50 (.. panic on duplicate read)

Does it work in a case like this?

nvartolomei commented:
> @bharathv: What err is consumerGroupReadInner expected to return if there was a duplicate read?

[It will panic](https://github.com/redpanda-data/redpanda/blob/4ba1a679a54dabb80cb9ab580919c69e996d39e7/tests/go/kgo-verifier/pkg/worker/verifier/validator_status.go#L101) rather than return.

> @bharathv: Does it work in a case like this?

Thanks for detailed explanation. No, the current logic does not account for this. I didn't consider this case. Considered only the cases where a broker would die and next PollFetches returns an error which we propagate and then ResetMonotonicityOffsets for.

May I suggest that instead of the implementation in this PR we instead do something like resetting the "monotonicity validator state" for partition via franz-go callbacks like 

https://github.com/twmb/franz-go/blob/1eb651c40a997fbbe35c7da19686c57cd65ba352/pkg/kgo/config.go#L1798-L1816

bharathv committed:
kgo/group_reader: revisit monotonicity state tracking

monotonicity state is now tracked based on group consumer hooks in
franz-go.

OnPartitionAssigned - assigns partition to a consumer in the group
OnOffsetsFetched - fetches offsets for an assigned partition
OnPartitionRevoked - revokes partition assignment from the group.

Since rebalance is async, we use BlockRebalaneOnPoll to avoid race
conditions between (for example) revoking a partition and processing
polled records.

Additionally redefines exactly once configuration described in the
previous commits.

Renamed to --group-consume-exactly-once and the scope is limited to
enforcing duplicate detection only when consuming with a consumer group.
Non consumer group reads are expected to be exactly once by default.

bharathv commented:
> > @bharathv: What err is consumerGroupReadInner expected to return if there was a duplicate read?
> 
> [It will panic](https://github.com/redpanda-data/redpanda/blob/4ba1a679a54dabb80cb9ab580919c69e996d39e7/tests/go/kgo-verifier/pkg/worker/verifier/validator_status.go#L101) rather than return.
> 
> > @bharathv: Does it work in a case like this?
> 
> Thanks for detailed explanation. No, the current logic does not account for this. I didn't consider this case. Considered only the cases where a broker would die and next PollFetches returns an error which we propagate and then ResetMonotonicityOffsets for.
> 
> May I suggest that instead of the implementation in this PR we instead do something like resetting the "monotonicity validator state" for partition via franz-go callbacks like
> 
> https://github.com/twmb/franz-go/blob/1eb651c40a997fbbe35c7da19686c57cd65ba352/pkg/kgo/config.go#L1798-L1816

Ya I think thats a cleaner approach, I took a stab at it, will re-request reviews once the  CI is green.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77384-->
<!--4da7dc11-ab94-43b4-9577-242b2efec858-->
<!--44547575-fa68-4b86-bde5-21d849c608fa-->
<!--c4367bd3-fe4b-4e68-8c55-628aee85aeb7-->
<!--27d75bf1-af3c-43bf-976b-d5f39c1ac726-->
<!--c9e5fa72-b6f1-4533-8a40-994f4ac5a1c0-->
<!--4a18c2cc-1807-4b3d-8852-98570123a288-->
<!--title-->
## Retry command for [Build#77384](https://buildkite.com/redpanda/redpanda/builds/77384)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_consumer_groups_mirroring@{"source_cluster_spec":{"cluster_type":"redpanda"}}
tests/rptest/tests/consumer_group_recovery_tool_test.py::ConsumerOffsetsRecoveryToolTest.test_consumer_offsets_partition_count_change
tests/rptest/tests/consumer_group_recovery_test.py::ConsumerOffsetsRecoveryTest.test_consumer_offsets_partition_recovery@{"force_offset_upload_failures":false}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_consumer_groups_mirroring@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/consumer_group_recovery_test.py::ConsumerOffsetsRecoveryTest.test_consumer_offsets_partition_recovery@{"force_offset_upload_failures":true}
```
<!--/body-->

bharathv commented:
@nvartolomei ready for review

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77471-->
<!--48b2007d-d035-4bbf-ab81-0c6ae2f34217-->
<!--9eb65d89-481c-4c19-aaf9-6b6d4dbda3b8-->
<!--d09e4a0e-7c61-4e47-83e5-513c3e9b3490-->
<!--46387451-0874-48b5-b60e-af30aed7ea11-->
<!--title-->
## Retry command for [Build#77471](https://buildkite.com/redpanda/redpanda/builds/77471)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/partition_move_interruption_test.py::PartitionMoveInterruption.test_cancellations_interrupted_with_restarts@{"replication_factor":3}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_topic_delete@{"source_cluster_spec":{"cluster_type":"redpanda"}}
tests/rptest/tests/consumer_offsets_consistency_test.py::ConsumerOffsetsConsistencyTest.test_flipping_leadership
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_topic_delete@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
```
<!--/body-->

bharathv commented:
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
dt-repeat=50
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_with_restart

bharathv commented:
This PR found a bunch of duplicate reads with the group consumer that earlier implementation didn’t catch. I’m starting to think if the default exactly_once expectation should be set to `false`. With chaos, leader changes restarts and everything else going on in tests, it’s pretty hard to avoid duplicate reads without transactions (which most tests don't).

nvartolomei commented on tests/go/kgo-verifier/pkg/worker/verifier/group_read_worker.go:
```diff
@@ -259,54 +416,19 @@ func (grw *GroupReadWorker) consumerGroupReadInner(
 
+	// Commit any uncommitted offsets at the end of the read loop, ignoring any errors
+	// since we're shutting down anyway
+	_ = client.CommitUncommittedOffsets(context.Background())
```
I believe this is wrong. I.e. it will commit offsets which were not validated too. I.e. for which EachRecord wasn't executed. I think this should be removed.

nvartolomei commented on tests/go/kgo-verifier/pkg/worker/verifier/group_read_worker.go:
```diff
@@ -142,6 +143,92 @@ func (cgs *ConsumerGroupOffsets) AddRecord(ctx context.Context, r *kgo.Record) {
+					lastConsumed = state.lastOffsetConsumed
+				}
+				if isDuplicate {
+					if grw.Status.Validator.exactlyOnceGroupConsumption {
```
I'm having a hard time understanding the value of this.

With consumer groups we can't achieve exactly once processing anyway. Committing offsets might have failed for example and, I guess, there are other possible conditions.

Btw, why would franz-go fetch offsets for an already assigned group (state.assigned == true being a required condition for `isDuplicate` to possibly be `true`).

nvartolomei reviewed (commented):
Having a bit of hard time following.

We know one thing from first principles. Exactly once consumption from the point of view of kgo when using consumer groups is impossible.

The simplest proof for this is the following sequence (S1):
- consume offset 0—ok
- attempt to commit offset 0—err
- fetch committed offsets: None
- consume offset 0—duplicate read

The other sequence is the one @bharathv raised (S2):
- consume offset 0-ok
- partition revoked
- partition assigned
- fetch committed offsets: None
- consume offset 0—duplicate read

S1 was already handled (poorly) by resetting monotonicity state on error.

Now we want to also handle S2 as a legal case.

First, we shouldn't introduce a new flag for it because from first principles we know that it is always a possibility for it to happen. I.e. the opposite can never be true.

There is one exception. If we (kgo) take control over committing offsets and we block or crash/halt if offset commit can't make progress. We'd also need to block partition revocation in that case until offsets are committed. This is not possible in the general case but in the case of kgo I could see this being possible with a couple of locks across goroutines to ensure that partition assignment is rather blocked until the peer goroutine committed the offsets.

Anyhow, this is all rather complicated and likely unnecessary.

I believe, the only reasonable thing we can do is during `OnPartitionAssigned` to reset the monotonicity state for the particular partition.

(Possibly we also need to take care that the peer goroutine doesn't mess up with the partition state after that concurrently. I.e. the franz-go client where partition was revoked might have an inflight commit or something. I'm not certain about this case.)

Then, we can improve from there. I.e. instead of resetting on PartitionAssignment we can track succeeded offset commits and reset to the latest offset commit.

I believe that's the best we can do. We don't need special flags for this.

---

I realize now that there are possible concurrency issues with callbacks. `BlockRebalanceOnPoll` is called but that is in context of a single franz-go client. We have one per worker and all mutate the same validator state.

In theory, the following callback call sequence is valid:

worker-1: onPartitionAssigned
worker-2: onPartitionAssigned
worker-1: onPartitionRevoked

I.e. implementation is possibly subtly incorrect.

Isn't it?

---

@bharathv do we have different understanding of first principles? Is what I'm suggesting wrong? Too complicated? Incomplete?
